design_notes/XDN03 - xlang Type System.md

```diff
@@ -15,3 +15,142 @@ status: draft
+interface specifies the signatures for a group of object members (aka methods, properties and
+events) but does not specify an implementation for those object members.
+
+Interfaces can be implemented by multiple different classes. Each class implementation of a given
+interface's object members is unrelated to any other class's implementation of the same interface.
+
+Interfaces may optionally inherit from a single other interfaces. All of the type members of the
+base interface are considered part of the derived interface as well. Any class implementing an
+interface that uses inheritance must implement all of the type members specified by both the base
+and derived interfaces.
+
+> Note, WinRT interfaces support inheritance from multiple other interfaces. This approach leads to
+inefficiencies at the ABI layer that xlang wishes to avoid by constraining interfaces to single
+inheritance. This decision may be revisited as the xlang project progresses.
+
+Interfaces are not versionable. Once they have been defined, they may never be changed.
+
+### Parameterized Interfaces
+
+xlang interfaces support type parameterization. Parameterized interfaces specify one or more type
+parameters that can be used as parameter types or return values of type members.
+
+Parameterized interfaces can inherit from both parameterized and non-parameterized interfaces.
+When inheriting from a parameterized interface, the type parameters of the base interface must
+be specified, though they may be specified in terms of the derived interface's type parameters.
+
+Non-parameterized interfaces can only inherit from concrete interfaces. This includes
+non-parameterized interfaces as  well as parameterized interfaces where all of the type parameters
+have been fully specified.
+
+Array types may not be used as type parameters.
+
+#### Parameterized Interface Examples
+
+For example, xlang's foundation types includes a set of parameterized interfaces to represent
+collections.
+
+- `IIterable<T>` represents a collection that can be enumerated
+- `IVector<T>` represents a collection that can be randomly accessed
+- `IMap<K,V>` represents a collection that maps from keys to values
+
+Both `IVector<T>` and `IMap<K,V>` can be enumerated. `IVector<T>`'s derives from `IIterable<T>`.
+In this case, the same type parameter is used for both `IVector` and `IIterable`. However,
+`IMap<K,V>` derives from `IIterable<IKeyValuePair<K,V>>`. In this case, the `T` type parameter of
+`IIterable` is instantiated as parameterized type `IKeyValuePair<K,V>`, using the type parameters
+defined as part of of `IMap<K,V>`.
+
+Using the same collection interfaces defined above, you could define an non parameterized interface
+`IPropertySet` that inherits from `IMap<String, Object>`. Note the use of concrete types in the
+declaration of IMap, where in the previous example we used type parameters.
+
+## Delegates
+
+## Classes
```

Comment:
make sure we discuss interface member collisions (cowboy-artist), maybe mention explicit implementation as a projection language feature

Reply:
There will be an entire section on interface members that includes a section on overloading. There's a [similar section](https://github.com/Microsoft/xlang/blob/master/docs/XSPEC01%20-%20Type%20System%20Specification.md#method-overloading) in the WinRT Type system doc.